compiler/rustc_data_structures/src/profiling.rs

```diff
@@ -288,6 +288,72 @@ impl SelfProfilerRef {
         })
     }
 
+    /// Start profiling a generic activity, allowing costly arguments to be recorded. Profiling
+    /// continues until the `TimingGuard` returned from this call is dropped.
+    ///
+    /// If the arguments to a generic activity are cheap to create, use `generic_activity_with_arg`
+    /// or `generic_activity_with_args` for their simpler API. However, if they are costly or
+    /// require allocation in sufficiently hot contexts, then this allows for a closure to be called
+    /// only when arguments were asked to be recorded via `-Z self-profile-events=args`.
+    ///
+    /// In this case, the closure will be passed an `EventArgRecorder`, to help with recording one
+    /// or many arguments to the generic activity being profiled, by calling one of its `record`
+    /// methods.
+    ///
+    /// Note: calling either of these methods is *required* for the self-profiling to create the
+    /// `TimingGuard`. A panic will be triggered if that doesn't happen.
+    ///
+    /// We could silently ignore this case and just record an event with no additional data here,
+    /// but this function exists explicitly to record arguments, so it fails loudly when there are
+    /// no such arguments to record.
+    ///
+    #[inline(always)]
+    pub fn generic_activity_with_arg_recorder<F>(
+        &self,
+        event_label: &'static str,
+        f: F,
+    ) -> TimingGuard<'_>
+    where
+        F: FnOnce(EventArgRecorder<'_>),
+    {
+        // Ensure this event will only be recorded when self-profiling is turned on.
+        self.exec(EventFilter::GENERIC_ACTIVITIES, |profiler| {
+            let builder = EventIdBuilder::new(&profiler.profiler);
+            let event_label = profiler.get_or_alloc_cached_string(event_label);
+
+            // Ensure the closure to create event arguments will only be called when argument
+            // recording is turned on.
+            let event_id = if profiler.event_filter_mask.contains(EventFilter::FUNCTION_ARGS) {
+                let mut event_id = EventId::INVALID;
+
+                // Set-up the builder, and call the user-provided closure to record potentially
+                // costly event arguments.
+                let recorder =
+                    EventArgRecorder { event_label, builder, profiler, event_id: &mut event_id };
+                f(recorder);
+
+                // It is expected that the closure will call one of the recorder's methods, in turn
+                // creating the event id we need for profiling. If that doesn't happen, it's a bug:
+                // the event id will still be invalid here.
+                //
+                // We could silently ignore this case and just record an event without arguments here,
+                // but we've been explicitly called in order to record arguments, so we fail loudly
+                // when there are no arguments to record.
+                if event_id == EventId::INVALID {
```

Comment:
This seems like the correct thing to do to me. Unless we discover there is measurable overhead to checking this (seems unlikely) or there is some compelling use case to call this but only sometimes record an argument, I think this is the way to go 👍